File: backend-python/app/models/submit_request.py
from sqlalchemy import Column, Integer, String, Text, Boolean, DateTime, ForeignKey
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from ..database import Base
import enum

class ProcessStatus(str, enum.Enum):
    WAITING = "WAITING"
    CONFIRMED = "CONFIRMED"
    APPROVED = "APPROVED"
    CANCEL = "CANCEL"
    REJECT = "REJECT"

class SubmitRequest(Base):
    __tablename__ = "submit_requests"

    id = Column(Integer, primary_key=True, index=True)
    account_name = Column(String, nullable=False)
    descriptions = Column(Text, nullable=False)
    supervisor_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    approver = Column(String, nullable=False)
    category_id = Column(Integer, ForeignKey("categories.id"), nullable=False)
    has_notify_mail = Column(Boolean, default=False)
    supervisor_approved = Column(Boolean, default=False)
    approver_approved = Column(Boolean, default=False)
    status = Column(Enum(ProcessStatus), default=ProcessStatus.WAITING)
    created_by_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    # Relationships
    supervisor = relationship("User", foreign_keys=[supervisor_id], backref="supervised_requests")
    created_by = relationship("User", foreign_keys=[created_by_id], backref="created_requests")
    category = relationship("Category", backref="submit_requests")

    def to_dict(self):
        return {
            "id": self.id,
            "account_name": self.account_name,
            "descriptions": self.descriptions,
            "supervisor_id": self.supervisor_id,
            "approver": self.approver,
            "category_id": self.category_id,
            "has_notify_mail": self.has_notify_mail,
            "supervisor_approved": self.supervisor_approved,
            "approver_approved": self.approver_approved,
            "status": self.status,
            "created_by_id": self.created_by_id,
            "created_at": self.created_at,
            "updated_at": self.updated_at,
        }

File: backend-java/src/main/java/com/crawldataai/backend/entity/SubmitRequest.java
package com.crawldataai.backend.entity;

import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;

@Entity
@Table(name = "submit_requests")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class SubmitRequest {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String accountName;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String descriptions;

    @ManyToOne
    @JoinColumn(name = "supervisor_id", referencedColumnName = "id")
    private User supervisor;

    @Column(nullable = false)
    private String approver;

    @ManyToOne
    @JoinColumn(name = "category_id", nullable = false)
    private Category category;

    private boolean hasNotifyMail = false;

    private boolean supervisorApproved = false;

    private boolean approverApproved = false;

    @Enumerated(EnumType.STRING)
    private ProcessStatus status = ProcessStatus.WAITING;

    @ManyToOne
    @JoinColumn(name = "created_by", referencedColumnName = "id")
    private User createdBy;

    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    public enum ProcessStatus {
        WAITING, APPROVED, CONFIRMED, REJECT, CANCEL, PROCESSING, COMPLETED, FAILED
    }
}
